File: leetcode/string/214. Shortest Palindrome/js/main.js
/*
Мы можем найти самый большой сегмент с нуля до i, который является палиндромом, и затем мы можем легко перевернуть оставшийся сегмент и добавить к началу.

Строка "abcba bcab".

Здесь самый большой палиндромный сегмент от начала - «abcba», а оставшийся сегмент - «bcab».

Следовательно, необходимая строка обратна «bcab» (= «bacb») + исходная строка (= «abcbabcab») = «bacb abcba» bcab».

Пример:
  s = dedcba -> abcded

  dedcba

  abcded - none
  bcded - none
  cded - none
  ded - true

  Возвращаем abc + ded + cba

  abcd

  dcba
  cba
  ba
  a
*/

// Time O(N^2)
// Space O(N)
const shortestPalindrome = str => {
  let rev = '';

  for (let i = str.length - 1; i >= 0; i--) {
    rev += str[i];
  }

  for (let i = 0; i < str.length; i++) {
    if (str.startsWith(rev.substring(i))) {
      return rev.substring(0, i) + str;
    }
  }

  return '';
};

File: leetcode/string/214. Shortest Palindrome/py/main.py
# Algo steps:

# The main point here found the largest palindrome substring starting from the first character
# Then reverse the remaining part as the prefix to s

class Solution:
    # Time O(N^2)
    # Space O(N)
    def shortestPalindrome(self, s):
        rev = s[::-1]
        
        for i in range(len(s)):
            if s.startswith(rev[i:]):
                return rev[:i] + s
        return s
    
    # Time O(N^2)
    # Space O(N)
    def shortestPalindrome_II(self, s):
        # largest palindrome starting from the first char
        def getLongestPalindrome(s):
          n = len(s)
          
          for endIndex in reversed(range(n)):
              left, right = 0, endIndex

              while left <= right and s[left] == s[right]:
                  left += 1
                  right -= 1

              left -= 1
              right += 1

              if left == right:
                  return endIndex

              return -1

        endIndex = getLongestPalindrome(s)

        prefix = s[endIndex+1:]

        return prefix + s
